### DISCUSSION TITLE
Check for object ownership - best practices

### DISCUSSION THREAD

**Discussion by florianmartens** [prisma/prisma#7147] [Q&A] [answered]
I'm new to working with Prisma. One aspect that is unclear to me is the right way to check if a user has permission on an object. Let's assume we have Book and Author models. Every book has an author (one-to-many). Only authors have permission to delete books.

An easy way to enforce this would be this:

```ts
prismaClient.book.deleteMany({
    id: bookId, <-- id is known
    author: {
        id: userId <-- id is known
    }
})
```
But this way it's very hard to show an UnauthorizedError to the user. Instead, the response will be a 500 status code since we can't know the exact reason why the query failed.

The other approach would be to query the book first and check the author of the book instance, which would result in one more query.

Is there a best practice for this in Prisma?

**Comment by ryands17** [comment]
@florianmartens 👋 

> But this way it's very hard to show an UnauthorizedError to the user.

Could you elaborate a bit on why it would be hard to show an error in this case? You would need to add a `try/catch` mechanism for this where you check if the deleted rows count is **0** and then you can return an `UnauthorizedError` back as the response.

**Comment by florianmartens** [comment]
Hi @ryands17,

maybe I phrased it a little too drastic. If I'm not mistaken, the count would also be 0 if there was no record with the given id. So differentiating between `404` and `401` would be hard. But additionally, we have to apply these patterns to all sorts of requests Get, Put, Post, Delete.

Starting from the example above: Let's say only book authors are allowed to retrieve a list of books. I'm performing this query to enforce authorization:

```ts
prismaClient.book.findMany({
    author: {
        id: userId <-- id is known
    }
})
```
If the result is an empty array this may mean that the author does not have any books. It may also mean that the user does not have permission to retrieve the list. Similarly, all API operations come with their own logic that may imply that a user lacks permission. This may end up being quite verbose and repetitive. I was wondering if best practices exist around this?

**Reply by ryands17** [reply, marked as answer]
In this case then, what you need are [middleware handlers](https://expressjs.com/en/guide/using-middleware.html) for your API. For e.g. Express has middleware where you check if the request is valid or not. In this case, what you can do is add a middleware to check if the user really has access to the book and then return an `UnauthorizedError` from there itself. This will not be required for a `findMany` and just for finding a single book, updating, and deleting. 

If the user indeed has the book, then the middleware will succeed and that would go to the actual logic of performing the operation. So in this case, you would require some sort of handler before the API operation that checks this. This is how I would personally do it.

**Reply by florianmartens** [reply]
Hi @ryands17, ok thanks!

Adding middleware will basically do what I mentioned above: Do an additional query to the database to retrieve the book and check the authors on it. But it's good to know that this is the default way to do this and that Prisma does not have baked-in way of doing this. I just wanted to be sure, I'm not missing anything.

Thanks!

**Reply by ryands17** [reply]
Yes the default way of doing this would be in your application logic itself as Prisma currently doesn't have any in-built method for this 👍
